pkg/app/provider/wopi/wopi.go

```diff
@@ -0,0 +1,317 @@
+	// q.Add("folderurl", "...")
+	u, ok := user.ContextGetUser(ctx)
+	if ok { // else defaults to "Anonymous Guest"
+		q.Add("username", u.Username)
+	}
+
+	q.Add("appname", app)
+	q.Add("appurl", p.appURLs["edit"][ext])
+
+	if p.conf.AppIntURL != "" {
+		q.Add("appinturl", p.conf.AppIntURL)
+	}
+	if viewExts, ok := p.appURLs["view"]; ok {
+		if url, ok := viewExts[ext]; ok {
+			q.Add("appviewurl", url)
+		}
+	}
+	if p.conf.AppAPIKey != "" {
+		httpReq.Header.Set("ApiKey", p.conf.AppAPIKey)
+	}
+	if p.conf.IOPSecret == "" {
+		p.conf.IOPSecret = os.Getenv("REVA_APPPROVIDER_IOPSECRET")
+	}
+	httpReq.URL.RawQuery = q.Encode()
+
+	httpReq.Header.Set("Authorization", "Bearer "+p.conf.IOPSecret)
+	httpReq.Header.Set("TokenHeader", token)
+
+	openRes, err := p.wopiClient.Do(httpReq)
+	if err != nil {
+		return "", errors.Wrap(err, "wopi: error performing open request to WOPI server")
+	}
+	defer openRes.Body.Close()
+
+	if openRes.StatusCode != http.StatusFound {
+		return "", errors.Wrap(err, "wopi: unexpected status from WOPI server: "+openRes.Status)
+	}
+	appFullURL := openRes.Header.Get("Location")
+
+	log.Info().Msg(fmt.Sprintf("wopi: returning app URL %s", appFullURL))
+	return appFullURL, nil
+}
+
+func getAppURLs(c *config) (map[string]map[string]string, error) {
+	// Initialize WOPI URLs by discovery
+	httpcl := rhttp.GetHTTPClient(
+		rhttp.Timeout(time.Duration(5 * int64(time.Second))),
+	)
+
+	appurl, err := url.Parse(c.AppIntURL)
+	if err != nil {
+		return nil, err
+	}
+	appurl.Path = path.Join(appurl.Path, "/hosting/discovery")
+
+	discReq, err := http.NewRequest("GET", appurl.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	discRes, err := httpcl.Do(discReq)
+	if err != nil {
+		return nil, err
+	}
+	defer discRes.Body.Close()
+
+	var appURLs map[string]map[string]string
+
+	if discRes.StatusCode == http.StatusNotFound {
+		// this may be a bridge-supported app
+		discReq, err = http.NewRequest("GET", c.AppIntURL, nil)
+		if err != nil {
+			return nil, err
+		}
+		discRes, err = httpcl.Do(discReq)
+		if err != nil {
+			return nil, err
+		}
+		defer discRes.Body.Close()
+
+		buf := new(bytes.Buffer)
+		_, err = buf.ReadFrom(discRes.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		// scrape app's home page to find the appname
+		if !strings.Contains(buf.String(), c.AppName) {
+			// || (c.AppName != "CodiMD" && c.AppName != "Etherpad") {
+			return nil, errors.New("Application server at " + c.AppURL + " does not match this AppProvider for " + c.AppName)
+		}
+
+		// register the supported mimetypes in the AppRegistry: this is hardcoded for the time being
+		if c.AppName == "CodiMD" {
+			appURLs = getCodimdExtensions(c.AppURL)
+		} else if c.AppName == "Etherpad" {
+			appURLs = getEtherpadExtensions(c.AppURL)
+		}
+	} else if discRes.StatusCode == http.StatusOK {
+		var netZoneName string
+		if c.AppName == "Collabora" {
+			netZoneName = "external-http"
```

Comment:
I fear this is very CERN-deployment specific... we'll probably need to come up with something better

Reply:
Much better, thanks! I hope this is generic enough for all possible deployments ;-)